### ISSUE TITLE
Compatibility with ospalh's Add note id

### ISSUE THREAD

**Issue by mnhende2** [hgiesel/anki_add_note_id#2]
Greeting hgiesel,

I have been using ospalh's add note ID add-on for many, many years.  Because of that, all of my 20,000 plus cards all have ID's in their noteid field.

I have also share my decks with a lot of friends, and regularly update my cards, then reshare the decks, so that my friends can update their cards.  I assume this is all possible because they have the same Note ID. 

If I use your add-on, will it change all of my existing note ID's?  It sounds like you are using a "full" note ID and ospalh's was not.  But I don't know exactly the difference, and I don't know if that means my note IDs will change and thus would no longer make it so that I can share my deck with friends, for them to update their decks.  

 Thanks in advance for any help!

**Comment by hgiesel** [comment]
Yes, they would change.
ospalh's addon uses a note id truncated by the value 15e11.

However I could add an option to help you out, and add use the truncated note id instead.

**Comment by mnhende2** [comment]
Thanks for the prompt response and the willingness to help.  If you are able to add that option for backward compatibility, that would be great.

**Comment by hgiesel** [comment]
I pushed a new version which should address your issue. Have a look in the add-on description for details.

**Comment by mnhende2** [comment]
Thanks.  Super helpful.  I just left a thumbs up on the add-on page.  

One more question.  In your response to someone's comment in the add-on page you wrote:

> Your cards need to have a field that is identical to the field set in the settings. Inspect the settings going to "Tools > Add-Ons > Ze Add Note Id > Config".
> 
> For example, mine says:

```
{
"NoteIdFieldName": "noteid",
"ShowFullNoteId": true,
"ShowMenu": false
}
```

If I read that correctly, it appears the "NoteIdFieldName" used to be user adjustable in the config file.  That would be a nice feature, but I don't see that such an option when I open the config file.  Is that no longer available or am I looking in the wrong place?  My field name is "Note ID" and I wonder if changing the field name in my deck will make it so that when someone imports an updated deck of mine, if the different field names will create a problem to deck being properly updated.

**Comment by hgiesel** [comment]
This comment is for the old version of the add-on (pre v2.1.28).

You can name the field however you want to. Look at this image:

![image](https://gitlab.com/hengiesel/anki_add_note_id/-/raw/master/images/field_dialog.png)

You get there on there through "Tools > Manage Note Types... > Fields", or the "Fields..." button in the Browser. The important part is only that it's the first in the list (so it's used for detecting duplicates), and that the checkbox is checked. The name can be whatever you want.

You have to do this per model, so you can even have the note id inserted into fields with different names per model.

**Comment by mnhende2** [comment]
> The important part is only that it's the first in the list (so it's used for detecting duplicates), and that the checkbox is checked. The name can be whatever you want.

Thanks! Super helpful and much appreciated.

**Comment by mnhende2** [comment]
Sorry to be back again so quickly, but it didn't work.  My "Note ID" is the first field.  And I went into fields, and I clicked the box to "Insert note ID" and then saved it.  But then it changed all of my Note ID's.

For example, one of them went from `1290201785625` to `-178995974000`.  This happened on two different decks that I tested on.  And most of the numbers turned into negative numbers. 

So, it seems that it isn't preserving the original Note ID, and thus will result in the conflicts I wrote about earlier.

I am going to restore from backup to get the Note ID's restored, but any other ideas?

**Comment by hgiesel** [comment]
So you have IDs in this field, which are not actually the note ID of the note (but rather used to be)?
Do I understand you correctly?

For example could you check what's actually the note id of the example you noted above (1290201785625)? By clicking Ctrl-Shift-I?

**Comment by mnhende2** [comment]
Correct.  All of the "Note ID" fields were populated. 20,000+ cards of them all populated.  Then when I ran your add-on and clicked the "Insert Note ID" box and saved it, it then overwrote all of the previous Note ID numbers.  And it replaced the numbers with usually a large negative number that is not the Note ID nor Card ID number of the card.  

You can see in the picture below. 

![Capture](https://user-images.githubusercontent.com/4404883/88941055-12b54700-d278-11ea-87bf-f36f8f071ef2.JPG)

**Comment by hgiesel** [comment]
I looked a bit through the code of old Add note id, and came across this:

https://github.com/ospalh/anki-addons/commit/96087164875ecd91019ecfc91886d6dcc109435e#diff-c93133cc4bf6bded39f945c20d7971f0

I didn't know that he changed this constant value the note id is decreased by over the years.
I can easily support this, however in a different way, rather than using a shortened note id, I will simply don't touch the note id on notes which already have a value.

It's obvious it's more important to you that the note id is unique, not that it corresponds to the actual note id.

**Comment by mnhende2** [comment]
@hgiesel:  Deeply and sincerely thanks so much for your help.   I restored my collection and then was going through dozens of sample cards to compare the "Note ID" listed in the field, and the actual "Note ID" of the card (as shown in Ctrl + Shift + 1).  I was terrified to find out that the numbers often do not match.  I don't know what ospalh did in his code to fill the Note ID field, but it clearly wasn't a direct one-to-one matching.  So, I had this sick feeling in my stomach that I have a massive problem across more than 20,000 cards.  😨 😭 

I was getting back on here to message you about my sad discovery only to find out that you have already thought of a brilliant solution for it.  Thanks again for the help, and the simple solution of ignoring notes that already have a value in the field!

**Comment by hgiesel** [comment]
To explain it shortly, he subtracted a constant value from it. In the latest version, it's 15e11, so 1500000000000. He made it for aesthetical reasons.

I approached him about this a year ago or so, but he didn't want to change it. This was the initial reason for me to create my own version, because I needed the note id in combination with Anki Connect.

Anyway, you can now try the new version I uploaded to Ankiweb.
In this new version, you go to the Addon configs, again and this time simply check the box for "Do not Overwrite note ID".

**Comment by mnhende2** [comment]
So hopefully one final question.  I have downloaded the new add-on, and I have gone into the config file and clicked the first box "Do not overwrite note IDs."  

I assume, I still go into fields and re-select the `Insert Note ID (overwrites current content!)` box.  Even though it says "overwrites current content", I assume that is negated from the setting in the config file, but just want to make sure that I'm not undoing the config setting.

**Comment by mnhende2** [comment]
Sorry to waste your time with that last question.  I realized that it was probably more helpful for me to just test it and report back to you.  So, the results are that checking the `Insert Note ID (overwrites current content!)` box in Fields does not overwrite current content, if the "do not overwrite" box is checked in the config settings.

Thanks again for all your help.  👏 🥳
